### PR TITLE
[FLINK-9891] Added hook to shutdown cluster if a session was created in per-job mode.

### PR DESCRIPTION
**Pull request by packet23**
## What is the purpose of the change

Change minimizes probability of a per-job yarn session cluster surviving after client exited. It restores the Flink 1.4 and below cli behaviour.

## Brief change log

  - when per-job yarn cluster is deployed, a shutdown hook installed
  - when shutdown hook is called, it terminates the deployed cluster
  - when cli terminates normally, shutdown hook is called by cli and deinstalled
  - otherwise, java runtime will call shutdown hook

## Verifying this change

We verified the change manually by submitting a DataSet API application to Yarn and killing client before job terminated.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 1]
```diff
@@ -1219,4 +1231,21 @@ public static void setJobManagerAddressInConfig(Configuration config, InetSocket
 		return constructor.newInstance(params);
 	}
 
+	private static class ClusterShutdownHook extends Thread {
+		private final ClusterClient<?> client;
+
+		ClusterShutdownHook(ClusterClient<?> client) {
```
Comment (yanghua) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: here use `@Nonnull` looks better to me

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 2]
```diff
@@ -249,13 +250,24 @@ protected void run(String[] args) throws Exception {
...
 				} else {
 					// also in job mode we have to deploy a session cluster because the job
 					// might consist of multiple parts (e.g. when using collect)
 					final ClusterSpecification clusterSpecification = customCommandLine.getClusterSpecification(commandLine);
 					client = clusterDescriptor.deploySessionCluster(clusterSpecification);
+					// if not running in detached mode, add a shutdown hook to shut down cluster if client exits
+					// there's a race-condition here if cli is killed before shutdown hook is installed
+					if (!runOptions.getDetachedMode()) {
+						shutdownHook = new ClusterShutdownHook(client);
+						Runtime.getRuntime().addShutdownHook(shutdownHook);
```
Comment (yanghua) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: In flink-core there is a `ShutdownHookUtil` util class you can consider

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java (hunk 1 above)
Comment (packet23) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: Thanks @yanghua for reviewing. Do you mean @Nonnull in addition to Objects.requireNonNull() or as replacement?

Comment (yanghua) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java]: Not replace, keep both of them.

**Comment by tillrohrmann** [comment]
I like the idea of making this behaviour configurable. @packet23 do you have time to add this functionality?

**Comment by GJL** [comment]
LGTM, merging.

**Pull request closed without merging**